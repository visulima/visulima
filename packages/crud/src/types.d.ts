import type { Handler as CreateHandler } from "./handler/create";
import type { Handler as DeleteHandler } from "./handler/delete";
import type { Handler as ListHandler } from "./handler/list";
import type { Handler as GetHandler } from "./handler/read";
import type { Handler as UpdateHandler } from "./handler/update";

export enum RouteType {
    CREATE = "CREATE",
    READ_ALL = "READ_ALL",
    READ_ONE = "READ_ONE",
    UPDATE = "UPDATE",
    DELETE = "DELETE",
}

export type ModelOption = {
    name?: string
    only?: RouteType[]
    exclude?: RouteType[]
    formatResourceId?: (resourceId: string) => string | number
};

export type ModelsOptions<M extends string = string> = {
    [key in M]?: ModelOption
};

export type HandlerOptions<M extends string = string> = {
    formatResourceId?: (resourceId: string) => string | number;
    models?: ModelsOptions<M>;
    exposeStrategy?: "all" | "none";
    pagination?: PaginationConfig,
    handlers?: {
        create?: CreateHandler;
        delete?: DeleteHandler;
        get?: GetHandler;
        list?: ListHandler;
        update?: UpdateHandler;
    },
};

export type PaginationConfig = {
    perPage: number
};

export interface HandlerParameters<T, Q> {
    adapter: Adapter<T, Q>;
    query: Q;
    resourceName: string;
}

export interface UniqueResourceHandlerParameters<T, Q> {
    adapter: Adapter<T, Q>;
    query: Q;
    resourceName: string;
    resourceId: string | number;
}

export interface Adapter<T, Q, M extends string = string> {
    models?: M[];
    init?: () => Promise<void>;
    parseQuery(resourceName: M, query: ParsedQueryParameters): Q;
    getAll(resourceName: M, query: Q): Promise<T[]>;
    getOne(resourceName: M, resourceId: string | number, query: Q): Promise<T>;
    create(resourceName: M, data: any, query: Q): Promise<T>;
    update(resourceName: M, resourceId: string | number, data: any, query: Q): Promise<T>;
    delete(resourceName: M, resourceId: string | number, query: Q): Promise<T>;
    getPaginationData(resourceName: M, query: Q): Promise<PaginationData>;
    getModels(): M[];
    connect?: () => Promise<void>;
    disconnect?: () => Promise<void>;
    handleError?: (error: Error) => void;
    mapModelsToRouteNames?: () => Promise<{ [key in M]?: string }>;
}

export type PaginationData = {
    total: number
    pageCount: number
    page: number
};

export type RecursiveField = {
    [key: string]: boolean | TRecursiveField;
};

export type WhereOperator = "$eq" | "$neq" | "$in" | "$notin" | "$lt" | "$lte" | "$gt" | "$gte" | "$cont" | "$starts" | "$ends" | "$isnull";

export type SearchCondition = string | boolean | number | Date | null;

export type WhereCondition = {
    [key in TWhereOperator]?: TSearchCondition;
};

export type Condition = {
    [key: string]: TSearchCondition | TWhereCondition | TCondition;
};

export type WhereField = Condition & {
    $and?: TCondition | TCondition[];
    $or?: TCondition | TCondition[];
    $not?: TCondition | TCondition[];
};

export type OrderByOperator = "$asc" | "$desc";

export type OrderByField = {
    [key: string]: TOrderByOperator;
};

export interface ParsedQueryParameters {
    select?: RecursiveField;
    include?: RecursiveField;
    where?: WhereField;
    orderBy?: OrderByField;
    limit?: number;
    skip?: number;
    distinct?: string;
    page?: number;
    originalQuery?: {
        [key: string]: any;
    };
}

export interface ExecuteHandler<Request, Response> {
    (request: Request, response: Response): Promise<void>;
}
