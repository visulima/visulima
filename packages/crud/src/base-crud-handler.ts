import type { HttpError } from "http-errors";
import createHttpError from "http-errors";
// eslint-disable-next-line import/no-extraneous-dependencies
import { ApiError } from "next/dist/server/api-utils";
import type { IncomingMessage, ServerResponse } from "node:http";

import createHandler from "./handler/create";
import deleteHandler from "./handler/delete";
import listHandler from "./handler/list";
import readHandler from "./handler/read";
import updateHandler from "./handler/update";
import parseQuery from "./query-parser";
import type { Adapter, ExecuteHandler, HandlerOptions, HandlerParameters, ParsedQueryParameters } from "./types.d";
import { RouteType } from "./types.d";
import formatResourceId from "./utils/format-resource-id";
import getAccessibleRoutes from "./utils/get-accessible-routes";
import { getResourceNameFromUrl } from "./utils/get-resource-name-from-url";
import getRouteType from "./utils/get-route-type";
import validateAdapterMethods from "./utils/validate-adapter-methods";

interface ResponseConfig {
    data: any;
    status: number;
}

async function baseHandler<R extends Request, Context, T, Q extends ParsedQueryParameters = any, M extends string = string>(
    responseExecutor: (responseOrContext: Context, responseConfig: ResponseConfig) => Promise<Response>,
    finalExecutor: (responseOrContext: Context) => Promise<void>,
    adapter: Adapter<T, Q>,
    options?: HandlerOptions<M>,
): Promise<ExecuteHandler<R, Context>>;

async function baseHandler<R extends IncomingMessage, RResponse extends ServerResponse, T, Q extends ParsedQueryParameters = any, M extends string = string>(
    responseExecutor: (responseOrContext: RResponse, responseConfig: ResponseConfig) => Promise<void>,
    finalExecutor: (responseOrContext: RResponse) => Promise<void>,
    adapter: Adapter<T, Q>,
    options?: HandlerOptions<M>,
): Promise<ExecuteHandler<R, RResponse>>;

// eslint-disable-next-line sonarjs/cognitive-complexity,func-style
async function baseHandler<R extends { method: string; url: string }, RResponse, T, Q extends ParsedQueryParameters = any, M extends string = string>(
    responseExecutor: (responseOrContext: RResponse, responseConfig: ResponseConfig) => Promise<RResponse>,
    finalExecutor: (responseOrContext: RResponse) => Promise<void>,
    adapter: Adapter<T, Q>,
    options?: HandlerOptions<M>,
): Promise<ExecuteHandler<R, RResponse>> {
    try {
        validateAdapterMethods(adapter);
    } catch (error_: any) {
        const error = error_ as HttpError;

        throw new ApiError(error.statusCode, error.message);
    }

    await adapter.init?.();

    const config = {
        formatResourceId,
        pagination: {
            perPage: 20,
        },
        ...options,
    };

    const routeNames = await adapter.mapModelsToRouteNames?.();
    const modelRoutes: { [key in M]?: string } = {};

    adapter.getModels().forEach((modelName) => {
        modelRoutes[modelName as M] = config.models?.[modelName as M]?.name ?? routeNames?.[modelName] ?? modelName;
    });

    return async (request, responseOrContext) => {
        const { modelName, resourceName } = getResourceNameFromUrl(request.url, modelRoutes as { [key in M]: string });

        if (!resourceName) {
            if (process.env.NODE_ENV === "development") {
                const mappedModels = await adapter.mapModelsToRouteNames?.();

                if (typeof mappedModels === "object") {
                    throw createHttpError(404, `Resource not found, possible models: ${Object.values(mappedModels).join(", ")}`);
                }
            }

            throw createHttpError(404, `Resource not found: ${request.url}`);
        }

        const { resourceId, routeType } = getRouteType(request.method, request.url, resourceName);

        if (routeType === null) {
            throw createHttpError(404, `Route not found: ${request.url}`);
        }

        const modelConfig = options?.models?.[modelName as M];

        const accessibleRoutes = getAccessibleRoutes(modelConfig?.only, modelConfig?.exclude, options?.exposeStrategy ?? "all");

        if (!accessibleRoutes.includes(routeType)) {
            throw createHttpError(404, `Route not found: ${request.url}`);
        }

        try {
            const resourceIdFormatted = modelConfig?.formatResourceId?.(resourceId as string) ?? config.formatResourceId(resourceId as string);

            await adapter.connect?.();

            const parsedQuery = parseQuery(request.url);
            const parameters: HandlerParameters<T, Q> = {
                adapter,
                query: adapter.parseQuery(modelName as M, parsedQuery),
                resourceName: modelName,
            };

            try {
                let responseConfig: ResponseConfig;

                switch (routeType) {
                    case RouteType.READ_ONE: {
                        responseConfig = await (config.handlers?.get ?? readHandler)<T, Q>({
                            ...parameters,
                            resourceId: resourceIdFormatted,
                        });
                        break;
                    }
                    case RouteType.READ_ALL: {
                        responseConfig = await (config.handlers?.list ?? listHandler)<T, Q>({
                            ...parameters,
                            pagination: config.pagination,
                            query: {
                                ...parameters.query,
                                limit: parsedQuery.limit ? Number(parsedQuery.limit) : undefined,
                                page: parsedQuery.page ? Number(parsedQuery.page) : undefined,
                            },
                        });
                        break;
                    }
                    case RouteType.CREATE: {
                        responseConfig = await (config.handlers?.create ?? createHandler)<T, Q, R>({
                            ...parameters,
                            request: request as R & { body: Record<string, any> },
                        });
                        break;
                    }
                    case RouteType.UPDATE: {
                        responseConfig = await (config.handlers?.update ?? updateHandler)<T, Q, R>({
                            ...parameters,
                            request: request as R & { body: Partial<T> },
                            resourceId: resourceIdFormatted,
                        });
                        break;
                    }
                    case RouteType.DELETE: {
                        responseConfig = await (config.handlers?.delete ?? deleteHandler)<T, Q>({
                            ...parameters,
                            resourceId: resourceIdFormatted,
                        });
                        break;
                    }
                    default: {
                        responseConfig = {
                            data: "Method not found",
                            status: 404,
                        };
                    }
                }

                await responseExecutor(responseOrContext, responseConfig);
            } catch (error: any) {
                if (adapter.handleError && !(error instanceof ApiError)) {
                    adapter.handleError(error);
                } else {
                    throw error;
                }
            }
        } finally {
            await adapter.disconnect?.();

            await finalExecutor(responseOrContext);
        }
    };
}

export default baseHandler;
