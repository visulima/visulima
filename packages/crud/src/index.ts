export { default as PrismaAdapter } from "./adapter/prisma";

export { RouteType } from "./types.d";
export type {
    ParsedQueryParameters,
    PaginationConfig,
    HandlerParameters,
    Adapter,
    ModelsOptions,
    HandlerOptions as CrudHandlerOptions,
    ModelOption,
    UniqueResourceHandlerParameters,
    Condition,
    OrderByField,
    OrderByOperator,
    RecursiveField,
    WhereField,
    WhereOperator,
    WhereCondition,
    SearchCondition,
    PaginationData,
} from "./types.d";

export type { SwaggerModelsConfig } from "./swagger/types.d";

export type { ModelsToOpenApiParameters } from "./swagger/adapter/prisma";
export { default as modelsToOpenApi } from "./swagger/adapter/prisma";
