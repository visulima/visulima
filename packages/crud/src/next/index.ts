export { default as edgeHandler } from "./api/edge";
export { default as nodeHandler } from "./api/node";
