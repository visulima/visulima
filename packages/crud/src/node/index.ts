import type { IncomingMessage, ServerResponse } from "node:http";

import baseHandler from "../base-crud-handler";
import type {
    Adapter, ExecuteHandler, HandlerOptions, ParsedQueryParameters,
} from "../types.d";

async function handler<R extends IncomingMessage, Response extends ServerResponse, T, Q extends ParsedQueryParameters = any, M extends string = string>(
    adapter: Adapter<T, Q>,
    options?: HandlerOptions<M>,
): Promise<ExecuteHandler<R, Response>> {
    return baseHandler<R, Response, T, Q, M>(
        async (response, responseConfig) => {
            response.setHeader("content-type", "application/json; charset=utf-8");

            response.statusCode = responseConfig.status;
            response.end(JSON.stringify(responseConfig.data, null, 2));
        },
        async (response) => {
            (response as Response).end();
        },
        adapter,
        options,
    );
}

export default handler;
