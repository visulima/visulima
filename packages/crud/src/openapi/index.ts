export type { SwaggerModelsConfig } from "./types";
export { default as getModelsAccessibleRoutes } from "./utils/get-models-accessible-routes";
export { default as getSwaggerPaths } from "./utils/get-openapi-paths";
export { default as getSwaggerTags } from "./utils/get-openapi-tags";
