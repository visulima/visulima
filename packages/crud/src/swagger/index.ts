export type { SwaggerModelsConfig } from "./types.d";
export { default as getModelsAccessibleRoutes } from "./utils/get-models-accessible-routes";
export { default as getSwaggerPaths } from "./utils/get-swagger-paths";
export { default as getSwaggerTags } from "./utils/get-swagger-tags";
