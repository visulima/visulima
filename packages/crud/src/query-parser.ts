import { URL } from "node:url";

import { setProperty } from "dot-prop";

import type { OrderByField, ParsedQueryParameters, RecursiveField, WhereField } from "./types";

const parseRecursive = (select: string): RecursiveField => {
    const selectFields: RecursiveField = {};

    const fields = select.split(",");

    fields.forEach((field) => {
        setProperty(selectFields, field, true);
    });

    return selectFields;
};

const parseWhere = (where: string): WhereField => {
    const whereObject = JSON.parse(where);
    const parsed: WhereField = {};

    Object.keys(whereObject).forEach((key) => {
        setProperty(parsed, key, whereObject[key]);
    });

    return parsed;
};

const parseOrderBy = (orderBy: string): OrderByField => {
    const parsed: OrderByField = {};
    const orderByObject = JSON.parse(orderBy);

    if (Object.keys(orderByObject).length > 0) {
        const key = Object.keys(orderByObject)[0] as string;

        if (orderByObject[key as keyof typeof orderByObject] === "$asc" || orderByObject[key as keyof typeof orderByObject] === "$desc") {
            parsed[key] = orderByObject[key as keyof typeof orderByObject];
        }
    }

    if (Object.keys(parsed).length !== 1) {
        throw new Error("orderBy needs to be an object with exactly 1 property with either $asc or $desc value");
    }

    return parsed;
};

// eslint-disable-next-line sonarjs/cognitive-complexity
const parseQuery = (url?: string): ParsedQueryParameters => {
    if (url) {
        const { searchParams } = new URL(url);
        const parsedQuery: ParsedQueryParameters = {};

        if (searchParams.get("select")) {
            parsedQuery.select = parseRecursive(searchParams.get("select") as string);
        }

        if (searchParams.get("include")) {
            parsedQuery.include = parseRecursive(searchParams.get("include") as string);
        }

        if (searchParams.get("where")) {
            parsedQuery.where = parseWhere(searchParams.get("where") as string);
        }

        if (searchParams.get("orderBy")) {
            parsedQuery.orderBy = parseOrderBy(searchParams.get("orderBy") as string);
        }

        if (searchParams.has("limit")) {
            parsedQuery.limit = Number.isFinite(+(searchParams.get("limit") as string)) ? +(searchParams.get("limit") as string) : undefined;
        }

        if (searchParams.has("skip")) {
            parsedQuery.skip = Number.isFinite(+(searchParams.get("skip") as string)) ? +(searchParams.get("skip") as string) : undefined;
        }

        if (searchParams.get("distinct")) {
            parsedQuery.distinct = searchParams.get("distinct") as string;
        }

        if (searchParams.get("page")) {
            parsedQuery.page = Number.isFinite(+(searchParams.get("page") as string)) ? +(searchParams.get("page") as string) : undefined;
        }

        return {
            originalQuery: Object.fromEntries(searchParams.entries()),
            ...parsedQuery,
        };
    }

    return {};
};

export default parseQuery;
