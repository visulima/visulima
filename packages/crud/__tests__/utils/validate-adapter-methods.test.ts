// eslint-disable-next-line max-classes-per-file
import {
    describe, expect, it, vi,
} from "vitest";

import { PrismaAdapter } from "../../src";
import validateAdapterMethods from "../../src/utils/validate-adapter-methods";

// eslint-disable-next-line no-constructor-return
class InvalidAdapter {}

describe("validateAdapterMethods", () => {
    it("should not throw a error for a valid adapter", () => {
        expect(() => validateAdapterMethods(
            new PrismaAdapter({
                prismaClient: vi.mock("@prisma/client", () => {
                    return {
                        PrismaClient: class {
                            public constructor() {
                                // eslint-disable-next-line no-constructor-return
                                return {
                                    // eslint-disable-next-line compat/compat
                                    $connect: () => Promise.resolve(),
                                    // eslint-disable-next-line compat/compat
                                    $disconnect: () => Promise.resolve(),
                                };
                            }
                        },
                    };
                }),
            }),
        )).not.toThrowError();
    });

    // @TODO: Add test for every method
    it("should throw a error for a invalid adapter", () => {
        // @ts-expect-error
        expect(() => validateAdapterMethods(new InvalidAdapter())).toThrowErrorMatchingSnapshot();
    });
});
