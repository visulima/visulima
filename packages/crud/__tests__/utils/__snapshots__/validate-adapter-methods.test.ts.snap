// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`validateAdapterMethods > should throw a error for a invalid adapter 1`] = `"Adapter must implement the \\"create\\" method."`;
